Add Technical and style guide to the contribution guide (#2250)
kubeflow/katib: docs/developer-guide.md

make undeploy
```

## Technical and style guide

The following guidelines apply primarily to Katib, 

> but other projects like training-operator might also adhere to them. 

Reviewer: ```suggestion
but other projects like [Training Operator](https://github.com/kubeflow/training-operator) might also adhere to them. 
```

Author: It makes sense.

Author: Done.

> 

Reviewer: Should we make a section about Golang here since Katib has various components written in Go, Python, Typescript?

```suggestion

## Golang Development

```

Author: SGTM

Author: Done.

When coding:

> - Follow [effective go](https://go.dev/doc/effective_go) guidelines.

Reviewer: Maybe we should add that developer should use `gofmt` to format their code.

Author: It would be better to focus on mentioning only items that it is difficult to automatically detect in CI.
WDYT?

Author: It makes sense.
As an alternative, I would suggest using `make check`. WDYT?

https://github.com/kubeflow/katib/blob/46173463027e4fd2e604e25d7075b2b31a702049/Makefile#L31

Reviewer: Sure, that even better, thanks!

Author: Done.

Testing:

> - Use `cmp.Diff` instead of `reflect.Equal`, to provide useful comparisons.

Reviewer: Would be nice to provide link to `cmp.Diff`.

Author: That makes sense.

Author: Done.

> - Define test cases as maps instead of slices to avoid dependencies on te running order.

Reviewer: Maybe we should say that:
```suggestion
- Define test cases as maps instead of slices to avoid dependencies on the running order. Set the test case name in the map key.
```

Author: That’s a great clarification!

Author: Done.
